server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java

```diff
@@ -53,12 +54,35 @@ public void reset(BytesRef bytesRef) throws IOException {
         treeOffset = 0;
     }
 
+    public int getHighestDimension() throws IOException {
+        ShapeType shapeType = getShapeType();
+        switch (shapeType) {
+            case POINT:
+            case MULTIPOINT:
+                return 1;
+            case LINESTRING:
+            case LINEARRING:
+            case MULTILINESTRING:
+                return 2;
+            case POLYGON:
+            case MULTIPOLYGON:
+            case ENVELOPE:
+            case CIRCLE:
+                return 3;
+            case GEOMETRYCOLLECTION:
+                return input.readVInt();
```

Comment:
Maybe we do not need an extra byte here, maybe we can have three different Geometry collections depending on the highest dimension?

Reply:
are you suggesting there be GeometryCollection, GeometryCollection3, Geometryacollection2, Geometryacollection1   in ShapeType? Or should there be a new enum specific to the tree's types

Reply:
Something like that, I haven't made up my mind. I do not want to overcomplicate things either.

Reply:
I guess the question is which worlds are more complicated:

- extend ShapeType to include specific GeometryCollection types with additional information about the highest dimension. If no dimension specified, then it is unknown.
  - **cons:** This would require custom mapping for the hypothetical future `geo_shape_type` aggregation, which would would only be slightly ugly, but contained. It would also convolute the ShapeType to not map 1-to-1 to proper shapes
- create a new Enum dedicated for type.
  - **cons:** need switch statements on reading and writing for conversion
- keep the original ShapeType, but write an extra bit for the dimension
  - **cons:** takes an extra bit and extra CPU to read the extra dimensional information

Reply:
I think option 2 is the clearer one of all the options. Since this type is dedicated to centroid support, it will feel cleaner if the CentroidCalculator owns tracking this information since your latest change updated it to be a geometry-visitor, so it can get this information too

Reply:
In reality we only need the highest dimension of the shape, the type is an `extra` :)

I think is more accurate if we keep dimension 0 for points, 1 for lines and two for polygons.

Reply:
> I think is more accurate if we keep dimension 0 for points, 1 for lines and two for polygons.

agreed. after I started updating it today, I felt the same way 👍 

> In reality we only need the highest dimension of the shape, the type is an extra :)

I think if we're already going to keep this information, we might as well group the two 
so that we can enable the shape-type aggregations we spoke of.